### DISCUSSION TITLE
shape different between getText and getPixmap

### DISCUSSION THREAD

**Discussion by tangent2018** [pymupdf/PyMuPDF#1143] [Q&A] [answered]
First, I used getText(option='dict') to get a dict. This dict includes width and height and some text information of the page.
Second, I used getPixmap to get a image. Then, I found there are two differences between the dict and image.
1, The width and height of image sometimes are a litter bigger than dict. (perhaps one pixel)
2, The width and height of image is reversed to dict. (ex. image width 800, height 500, dict width 500, height 800)

My purpose is to match information from dict to image. I want to know how can I match the width and height from image to dict, it seems it should rotate 90 degrees cw or ccw, but I don`t know how to decide it.

Thank you, Jorj.

code:
doc = fitz.open(pdf_path)
for page in doc:
  dict = page.getText(option='dict')
  
  pm = page.getPixmap()
  PNGdata = pm.getPNGdata()
  byte_stream = io.BytesIO(PNGdata)
  image = np.array(Image.open(byte_stream))

**Comment by JorjMcKie** [comment, marked as answer]
There is an important difference, how page rotation is handled by functions dealing with coordinates of page elements and functions rendering the page to a pixmap:
* Rendering the page always reflects how the page looks like - as if you would see it using a PDF viewer application.
* Text searching and extraction, image bbox computation, as well as inserting text, images and other stuff is always based on coordinates of the **_unrotated_** page. There are several good reasons for this.

So to synchronize, you need to either rotate the pixmap (`prerotate()` method) or (temporarily) de-rotate the page before rendering.
You could of course also compute the rotated coordinates of anything provided by text extraction etc.: multiply points and bboxes by `page.rotation_matrix` and `page.derotation_matrix` accordingly.

**Reply by tangent2018** [reply]
Thank you. It helps.

**Comment by JorjMcKie** [comment]
>  1, The width and height of image sometimes are a litter bigger than dict. (perhaps one pixel)

This is caused by the fact, that an image must have integer width and height, whereas the page itself has float dimensions.

**Reply by tangent2018** [reply]
Thank you. It helps.
